python/dlr/tf_model.py

```diff
@@ -0,0 +1,181 @@
+# coding: utf-8
+import logging
+import os
+import tensorflow as tf
+
+# A prefix that will be prepended to the names in graph_def
+PREFIX = "import"
+
+
+def _load_frozen_graph(frozen_graph_file, device):
+    """
+    Load Frozen Graph
+
+    Parameters
+    ----------
+    frozen_graph_file : str
+        Full path to frozen graph (.pb file)
+    device : str
+        device type and id, (e.g. /cpu:0)
+
+    Returns
+    -------
+    out : Graph :py:class:`tf.Graph`
+    """
+    logging.info("Loading frozen graph: {}".format(frozen_graph_file))
+    with tf.device(device):
+        with tf.gfile.GFile(frozen_graph_file, 'rb') as f:
+            graph_def = tf.GraphDef()
+            graph_def.ParseFromString(f.read())
+        with tf.Graph().as_default() as graph:
+            tf.import_graph_def(graph_def, name=PREFIX)
+        return graph
+
+
+def _get_input_and_output_names(graph):
+    """
+    Get input and output tensor names
+
+    Parameters
+    ----------
+    graph : tf.Graph
+        Tensorflow graph
+
+    Returns
+    -------
+    input_tensor_names : List of input tensor names
+    output_tensor_names  : List of output tensor names
+    """
+    input_tensor_names = []
+    output_tensor_names = set()
+    op_prefix = PREFIX + "/"
+    for op in graph.get_operations():
+        if not op.name.startswith(op_prefix):
+            continue
+        if op.type == 'Placeholder' and op.inputs.__len__() == 0 and op.outputs.__len__() == 1:
+            input_tensor_names.append(op.outputs[0].name)
+        if op.outputs.__len__() == 1:
+            output_tensor_names.add(op.outputs[0].name)
+    for op in graph.get_operations():
+        for in_t in op.inputs:
+            if in_t.name in output_tensor_names:
+                output_tensor_names.remove(in_t.name)
+    # Sort list of output tensor names in order to get consistent output in run()
+    output_tensor_names = list(output_tensor_names)
+    output_tensor_names.sort()
+    return input_tensor_names, output_tensor_names
+
+
+class TFModelImpl:
+    """
+    TFModelImpl is a wrapper on top of tensorflow which implements DLRModel API
+
+    Parameters
+    ----------
+    frozen_graph_file : str
+        Full path to frozen graph (.pb file)
+    dev_type : str
+        Optional. Device type ('cpu' or 'gpu')
+    dev_id : int
+        Optional. Device ID
+    """
+    def __init__(self, frozen_graph_file, dev_type=None, dev_id=None):
+        if not os.path.exists(frozen_graph_file):
+            raise ValueError("frozen_graph_file %s doesn't exist" % frozen_graph_file)
+        self.frozen_graph_file = frozen_graph_file
```

Comment:
Is this intended to be accessed externally? If not, we'd better set it to be internal.

Reply:
renamed to `_frozen_graph_file`